core/src/com/tgf/twf/core/world/storage/HashMapInventory.java

```diff
@@ -41,25 +84,61 @@ public boolean store(final Inventory inventory) {
         return true;
     }
 
+    @Override
+    public boolean reserve(final ResourceType resourceType, final int quantity) {
+        if (quantity == 0) {
+            return true;
+        }
+        final Stock stock = stocks.get(resourceType);
+        if (stock == null) {
+            stocks.put(resourceType, Stock.ofReservedQuantity(quantity));
+        } else {
+            stock.reservedQuantity += quantity;
+        }
+        return true;
+    }
+
+    @Override
+    public boolean reserve(final Inventory inventory) {
+        inventory.getStoredResourceTypes().forEach(
```

Comment:
Should this not be an `allMatch` or similar? Is the caller responsible for preventing partial reservations?

Reply:
This is the Inventory so it doesn't have the concept of Capacity.
The Storage is responsible to perform access control.

Reply:
Storage

    /**
     * @param resources The quantity of resources to store, per resource type.
     * @return true if the storage have enough capacity in store to satisfy the offer.
     */
    public boolean canStore(final Inventory resources) {
        return resources.getStoredResourceTypes().stream()
                .allMatch(resourceType -> capacity.getRemainingCapacity(inventory, resourceType, Capacity.CapacityCountMode.EXCLUDE_RESERATIONS)
                        >= resources.getStoredQuantity(resourceType));
    }

    /**
     * @param resources The quantity of resources to store. {@link #canStore(Inventory)} should be called first to check if the
     *                  storage can accept the offer or not. {@link Capacity} overflow will throw an {@link IllegalStateException}.
     */
    public void store(final Inventory resources) {
        if (!inventory.store(resources)) {
            throw new IllegalStateException("Storage rejected storage of " + resources);
        }
    }